src/laser_publisher.cpp

```diff
@@ -0,0 +1,208 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * Author: Chad Rockey
+ */
+
+#include <string>
+#include <vector>
+
+#include "laser_proc/laser_publisher.hpp"
+
+namespace laser_proc
+{
+
+typedef sensor_msgs::msg::LaserScan (* PublishFunction)(
+  const sensor_msgs::msg::MultiEchoLaserScan &
+  msg);
+
+struct LaserPublisher::Impl
+{
+  Impl()
+  : unadvertised_(false)
+  {}
+
+  ~Impl()
+  {
+    shutdown();
+  }
+
+  bool isValid() const
+  {
+    return !unadvertised_;
+  }
+
+  void shutdown()
+  {
+    if (!unadvertised_) {
+      unadvertised_ = true;
+    }
+  }
+
+  rclcpp::Publisher<sensor_msgs::msg::MultiEchoLaserScan>::SharedPtr echo_pub_;
+  std::vector<rclcpp::Publisher<sensor_msgs::msg::LaserScan>::SharedPtr> pubs_;
+  std::vector<PublishFunction> functs_;
+  bool unadvertised_;
+};
+
+///< @TODO Make a static class that creates these
+LaserPublisher::LaserPublisher(
+  rclcpp::Node::SharedPtr & nh, uint32_t queue_size,
+  /* const ros::SubscriberStatusCallback& connect_cb,
+  const ros::SubscriberStatusCallback& disconnect_cb,
+  const ros::VoidPtr& tracked_object, bool latch, */
+  bool publish_echoes)
+: impl_(new Impl)
+{
+  if (publish_echoes) {
+    // impl_->echo_pub_ = nh->create_publisher<sensor_msgs::msg::MultiEchoLaserScan>(
+    // "echoes", queue_size, connect_cb, disconnect_cb, tracked_object, latch);
+    impl_->echo_pub_ = nh->create_publisher<sensor_msgs::msg::MultiEchoLaserScan>("echoes",
+        queue_size);
+  }
+  // impl_->pubs_.push_back(nh->create_publisher<sensor_msgs::msg::LaserScan>(
+  // "first", queue_size, connect_cb, disconnect_cb, tracked_object, latch));
+  impl_->pubs_.push_back(nh->create_publisher<sensor_msgs::msg::LaserScan>("first", queue_size));
+  impl_->functs_.push_back(laser_proc::LaserProc::getFirstScan);
+  // impl_->pubs_.push_back(nh->create_publisher<sensor_msgs::msg::LaserScan>(
+  // "last", queue_size, connect_cb, disconnect_cb, tracked_object, latch));
+  impl_->pubs_.push_back(nh->create_publisher<sensor_msgs::msg::LaserScan>("last", queue_size));
+  impl_->functs_.push_back(laser_proc::LaserProc::getLastScan);
+  // impl_->pubs_.push_back(nh->create_publisher<sensor_msgs::msg::LaserScan>(
+  // "most_intense", queue_size, connect_cb, disconnect_cb, tracked_object, latch));
+  impl_->pubs_.push_back(nh->create_publisher<sensor_msgs::msg::LaserScan>("most_intense",
+    queue_size));
+  impl_->functs_.push_back(laser_proc::LaserProc::getMostIntenseScan);
+}
+
+uint32_t LaserPublisher::getNumSubscribers() const
+{
+  // TODO(karsten1987): no way to do this in ros2 yet
```

Comment:
This should be possible now; see https://github.com/ros-drivers/velodyne/blob/ec9adec6a523ab705ee526b06141aa9ad8d43c4f/velodyne_pointcloud/src/conversions/convert.cpp#L150 for an example.